pkg/policy/repository.go

```diff
@@ -351,51 +353,121 @@ nextLabel:
 // This is just a helper function for unit testing.
 // TODO: this should be in a test_helpers.go file or something similar
 // so we can clearly delineate what helpers are for testing.
-func (p *Repository) Add(r api.Rule) (uint64, error) {
+func (p *Repository) Add(r api.Rule, localRuleConsumers []IdentityConsumer) (uint64, map[uint16]struct{}, error) {
 	p.Mutex.Lock()
 	defer p.Mutex.Unlock()
 
 	if err := r.Sanitize(); err != nil {
-		return p.revision, err
+		return p.revision, nil, err
 	}
 
 	newList := make([]*api.Rule, 1)
 	newList[0] = &r
-	return p.AddListLocked(newList), nil
+	_, rev := p.AddListLocked(newList)
+	return rev, map[uint16]struct{}{}, nil
+}
+
+// NewIDSet returns a new instance of an IDSet.
+func NewIDSet() *IDSet {
+	return &IDSet{
+		IDs: map[uint16]struct{}{},
+	}
+}
+
+type IdentityConsumer interface {
+	GetID16() uint16
+	GetSecurityIdentity() *identity.Identity
+	PolicyRevisionBumpEvent(rev uint64)
 }
 
 // AddListLocked inserts a rule into the policy repository with the repository already locked
 // Expects that the entire rule list has already been sanitized.
-func (p *Repository) AddListLocked(rules api.Rules) uint64 {
-	newList := make([]*rule, len(rules))
+func (p *Repository) AddListLocked(rules api.Rules) (ruleSlice, uint64) {
+
+	newList := make(ruleSlice, len(rules))
 	for i := range rules {
-		newList[i] = &rule{Rule: *rules[i]}
+		newRule := &rule{
+			Rule:     *rules[i],
+			metadata: newRuleMetadata(),
+		}
+		newList[i] = newRule
 	}
+
 	p.rules = append(p.rules, newList...)
 	p.revision++
 	metrics.PolicyCount.Add(float64(len(newList)))
 	metrics.PolicyRevision.Inc()
 
-	return p.revision
+	return newList, p.revision
+}
+
+// UpdateLocalConsumers updates the cache within each rule in the given repository
+// which specifies whether said rule selects said identity. Returns a wait group
+// which can be used to wait until all rules have had said caches updated.
+func (p *Repository) UpdateLocalConsumers(identifiers []IdentityConsumer) *sync.WaitGroup {
+	var policySelectionWG sync.WaitGroup
+	for _, identityConsumer := range identifiers {
+		policySelectionWG.Add(1)
+		go p.rules.analyzeWhetherRulesSelectEndpoint(identityConsumer, NewIDSet(), &policySelectionWG)
+	}
+	return &policySelectionWG
 }
 
-// AddList inserts a rule into the policy repository.
-func (p *Repository) AddList(rules api.Rules) uint64 {
+// RemoveIdentifierFromRuleCaches removes identifier from the processedConsumers
+// and localRuleConsumers sets in each rule within the repository. Returns a
+// sync.WaitGroup which can be waited on once all rules have been processed in
+// relation to the identifier.
+func (p *Repository) RemoveIdentifierFromRuleCaches(identifier uint16) *sync.WaitGroup {
```

Comment:
What's an `Identifier`? (Again I assume endpointID, but if it's endpointID then let's name the functions and variables to show this).

Reply:
Fixed!